android/src/com/mapswithme/maps/intent/Factory.java

```diff
@@ -781,6 +781,10 @@ String getUrl()
     {
       Utils.checkNotNull(url);
       mUrl = url;
+
+      Uri mUri = Uri.parse(mUrl);
+      String mBackUrl = mUri.getQueryParameter("backurl");
+      MwmApplication.get().setBackUrl(mBackUrl);
```

Comment:
it's not what I meant. I try to explain here in comment how I would do it.

1. Let's decide that MAPSME app is launched by deeplink. For Android, it may be **https://dlink.maps.me/launch?back_uri=<caller_app_uri>**. For IOS we may use **mapsme:://launch?back_uri=<caller_app_uri>**. As for IOS implementation please discuss it with @beloal. Below, I'll deep in Android implementation only.

2. So, we have https://dlink.maps.me/launch?back_uri=<caller_app_uri>. Then, you should implement DlinkIntentProcessor in Factory.java. As sample, you may see on already existing DlinkGuidesPageIntentProcessor. In your case, you don't need an introduction task, just return null. But you need to create LaunchMapTask that should implement MapTask interface.

3. And then, in LaunchMapTask.**run** method you have an MwmActivity object and you can put extra, **extra_back_uri** for example, in MwmActivity's intent (target.getIntent().putExtra(.., back_uri)).

4. Finally, when KEYCODE_ESCAPE is handled, you check whether the activity's intent contains **extra_back_uri** extra or not.  And if it contains our extra you retreive the back uri from it and start the intent.

If you have any questions please let me know. Also, if you push this code in you fork in a separate branch, I may do it for you in your branch. Keep in mind it please.

Reply:
Thanks for the advice.  I'll start looking into implementing following your advice.  I've also created a new branch "WunderLINQ" as you suggested.

Reply:
Currently, I'm using the following link to launch MAPSME from my Android app.  Is this correct?

Navigation:
mapsme://route?sll=lat,lon&saddr=Start&dll=destlat,destlon&daddr=Home&type=vehicle&back_uri=wunderlinq://

Return to Navigation:
mapsme://?back_uri=wunderlinq://

Reply:
@alexzatsepin forgot to tag you on my questions/reply.

Reply:
Hi @kconger!

Sorry for delayed answer.

1. As for deeplink, For Android app I would suggest to use **https://dlink.maps.me/route** instead of mapsme://route.
2. I didn't get why we need a separate deeplink for "Return to Navigation"? We can store "back_uri" when MAPSME is launched by first deeplink and use stored parameter for further usage, i.e. when ESCAPE is clicked, can't we?

Reply:
Hi @alexzatsepin 

The "Return to Navigation"  is a way for my app to send you back to MAPSME without interrupting current navigation in MAPSME.

For example. User leaves MAPSME with ESCAPE to go back to the WunderLINQ app to check tire pressure, then the USER wants to go back to MAPSME to see navigation again.

Reply:
Hi @alexzatsepin

Ok, I think I've implemented what you've suggested.  Please let me know how it looks.

Thanks

Reply:
Hi kconger,

I need to change your code a little bit. Please push your changes in the branch that name is different from "master", so I will be able to checkout it and make commits directly in your branch. Thanks.

Reply:
Hi @alexzatsepin,

Thanks for taking a look.  I pushed them into branch "WunderLINQ".

Keith

Reply:
In description of PR I see " mapsme:master from kconger:master" that your want to merge this code from your master branch. Can you change the branch (on WunderLINQ) in this PR?

Reply:
Hi alexzatsepin,

Looks like you can't change the source branch of the PR.  So I create a new PR: 12316 from the 'WunderLINQ' branch.